src/block_gmres.jl

```diff
@@ -0,0 +1,349 @@
+      if !restart && (inner_iter > mem)
+        for i = 1 : inner_iter
+          push!(R, S(undef, p, p))
+        end
+      end
+
+      # Continue the block-Arnoldi process.
+      P = NisI ? V[inner_iter] : solver.P
+      NisI || mulorldiv!(P, N, V[inner_iter], ldiv)  # P ← NVₖ
+      mul!(W, A, P)                                  # W ← ANVₖ
+      MisI || mulorldiv!(Q, M, W, ldiv)              # Q ← MANVₖ
+      for i = 1 : inner_iter
+        mul!(R[nr+i], V[i]', Q)       # Ψᵢₖ = Vᵢᴴ * Q
+        mul!(Q, V[i], R[nr+i], α, β)  # Q = Q - Vᵢ * Ψᵢₖ
+      end
+
+      # Reorthogonalization of the block-Krylov basis.
+      if reorthogonalization
+        for i = 1 : inner_iter
+          mul!(Ψtmp, V[i]', Q)       # Ψtmp = Vᵢᴴ * Q
+          mul!(Q, V[i], Ψtmp, α, β)  # Q = Q - Vᵢ * Ψtmp
+          R[nr+i] .+= Ψtmp
+        end
+      end
+
+      # Vₖ₊₁ and Ψₖ₊₁.ₖ are stored in Q and Ztmp. 
+      householder!(Q, Ztmp, τ)
+
+      # Update the QR factorization of Hₖ₊₁.ₖ.
+      # Apply previous Givens reflections Ωᵢ.
+      # [cᵢ  sᵢ] [ r̄ᵢ.ₖ ] = [ rᵢ.ₖ ]
+      # [s̄ᵢ -cᵢ] [rᵢ₊₁.ₖ]   [r̄ᵢ₊₁.ₖ]
+      for i = 1 : inner_iter-1
+        Rtmp      =      c[i]  * R[nr+i] + s[i] * R[nr+i+1]
+        R[nr+i+1] = conj(s[i]) * R[nr+i] - c[i] * R[nr+i+1]
+        R[nr+i]   = Rtmp
+      end
+
+      # Compute and apply current Givens reflection Ωₖ.
+      # [cₖ  sₖ] [ r̄ₖ.ₖ ] = [rₖ.ₖ]
+      # [s̄ₖ -cₖ] [hₖ₊₁.ₖ]   [ 0  ]
+      (c[inner_iter], s[inner_iter], R[nr+inner_iter]) = sym_givens(R[nr+inner_iter], Hbis)
+
+      # Update Zₖ = (Qₖ)ᴴΓE₁
+      ζₖ₊₁          = conj(s[inner_iter]) * z[inner_iter]
+      z[inner_iter] =      c[inner_iter]  * z[inner_iter]
+
+      # Update residual norm estimate.
+      # ‖ M(B - AXₖ) ‖_F = ‖Ztmp‖_F
+      RNorm = norm(Ztmp)
+      history && push!(RNorms, RNorm)
+
+      # Update the number of coefficients in Rₖ
+      nr = nr + inner_iter
+
+      # Update stopping criterion.
+      solved = RNorm ≤ ε
+      inner_tired = restart ? inner_iter ≥ min(mem, inner_itmax) : inner_iter ≥ inner_itmax
+      timer = time_ns() - start_time
+      overtimed = timer > timemax_ns
+      kdisplay(iter+inner_iter, verbose) && @printf("%5d  %5d  %7.1e  %.2fs\n", npass, iter+inner_iter, rNorm, ktimer(start_time))
+
+      # Compute Vₖ₊₁.
+      if !(solved || inner_tired || overtimed)
+        if !restart && (inner_iter ≥ mem)
+          push!(V, S(undef, n, p))
+          push!(Z, S(undef, p, p))
+        end
+        V[inner_iter+1] .= Q
+        Z[inner_iter+1] = Ztmp
+      end
+    end
+
+    # Compute Yₖ by solving RₖYₖ = Zₖ with a backward substitution by block.
+    Y = Z  # Yᵢ = Zᵢ
+    for i = inner_iter : -1 : 1
+      pos = nr + i - inner_iter         # position of Ψᵢ.ₖ
+      for j = inner_iter : -1 : i+1
+        mul!(Y[i], R[pos], Y[j], α, β)  # Yᵢ ← Yᵢ - ΨᵢⱼYⱼ
+        pos = pos - j + 1               # position of Ψᵢ.ⱼ₋₁
+      end
+      Y[i] = Y[i] \ R[pos]  # Yᵢ ← Yᵢ \ Ψᵢᵢ
```

Comment:
how about using `ldiv!` here, to avoid using the backslash operator?

Reply:
I will use `ldiv!` if I can, I just want to make it work and after I will optimize the code.